iree/compiler/Conversion/CodegenUtils/ForOpCanonicalization.cpp

```diff
@@ -132,15 +132,99 @@ class ForOpArgFolding final : public OpRewritePattern<scf::ForOp> {
+///
+/// Those loop-carried values will be lowered into SPIR-V local variables. This
+/// pattern allows packing f16 values into f32 variables tightly so that we can
+/// generate shader conformant SPIR-V.
+struct PackForOpInductionVarVector final : public OpRewritePattern<scf::ForOp> {
+  using OpRewritePattern<scf::ForOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(scf::ForOp forOp,
+                                PatternRewriter& rewriter) const override {
+    VectorType v8f16Type = VectorType::get({8}, rewriter.getF16Type());
+    VectorType v4f32Type = VectorType::get({4}, rewriter.getF32Type());
+
+    SmallVector<unsigned, 8> ivIndices;
+    for (auto it : llvm::enumerate(forOp.getRegionIterArgs())) {
+      if (it.value().getType() == v8f16Type) ivIndices.push_back(it.index());
+    }
+    if (ivIndices.empty()) return failure();
+
+    // Bit cast all init values from v8f16 to v4f32.
+    auto ivInitValues = llvm::to_vector<8>(forOp.getIterOperands());
+    for (unsigned index : ivIndices) {
+      Value oldValue = ivInitValues[index];
+      ivInitValues[index] = rewriter.create<vector::BitCastOp>(
+          oldValue.getLoc(), v4f32Type, oldValue);
+    }
+
+    // Create a new loop with the casted init values. This also creates
+    // induction variables with proper type.
+    auto newLoop = rewriter.create<scf::ForOp>(
+        forOp.getLoc(), forOp.lowerBound(), forOp.upperBound(), forOp.step(),
+        ivInitValues);
+
+    // Move all operations to the new for op. This also replaces block
+    // arguments. to the new block arguments.
+    rewriter.mergeBlocks(forOp.getBody(), newLoop.getBody(),
+                         newLoop.getBody()->getArguments());
+
+    // Bit cast induction variables back to the original type to fix uses.
+    rewriter.setInsertionPointToStart(newLoop.getBody());
+    for (unsigned index : ivIndices) {
+      Value newIv = newLoop.getRegionIterArgs()[index];
+      auto bitcastOp =
+          rewriter.create<vector::BitCastOp>(newIv.getLoc(), v8f16Type, newIv);
+      // Replace all uses of the new induction variable with a bitcast. We need
+      // to exclude the bitcast op itself given it also uses the induction
+      // variable.
+      SmallPtrSet<Operation*, 1> exceptions;
+      exceptions.insert(bitcastOp);
+      newIv.replaceAllUsesExcept(bitcastOp, exceptions);
+    }
+
+    auto yieldOp = cast<scf::YieldOp>(newLoop.getBody()->getTerminator());
+    auto ivRetValues = llvm::to_vector<8>(yieldOp.getOperands());
+
+    // Bit cast return values to the new type to fix yield.
+    rewriter.setInsertionPoint(yieldOp);
+    for (unsigned index : ivIndices) {
+      Value oldRet = ivRetValues[index];
+      ivRetValues[index] = rewriter.create<vector::BitCastOp>(
+          oldRet.getLoc(), v4f32Type, oldRet);
+    }
+    yieldOp->setOperands(ivRetValues);
+
+    SmallVector<Value, 8> forRetValues;
+    for (Value result : newLoop.getResults()) forRetValues.push_back(result);
```

Comment:
you can do `llvm::to_vector<8>(newLoop.getResults())`. Or
`SmallVector<Value, 8> forRetValues(newLoop.getResults().begin(), newLoop.getResults().end());`

Reply:
I cannot do that because `getResults` returns an array of `OpResult`s.